Making a HEAD request to an empty S3 object that is gzipped throws an exception

### Steps to reproduce
1. Run the sample project (steps are in the README) - https://github.com/jbaranski/aws-s3-head-bug
2. Receive exception:
```
java.util.concurrent.ExecutionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: java.io.EOFException
    at java.util.concurrent.CompletableFuture.reportGet (CompletableFuture.java:396)
    at java.util.concurrent.CompletableFuture.get (CompletableFuture.java:2073)
    at com.jeffbaranski.bugreport.Main.main (App.java:34)
    at org.codehaus.mojo.exec.ExecJavaMojo.doMain (ExecJavaMojo.java:385)
    at org.codehaus.mojo.exec.ExecJavaMojo.doExec (ExecJavaMojo.java:374)
    at org.codehaus.mojo.exec.ExecJavaMojo.lambda$execute$0 (ExecJavaMojo.java:296)
    at java.lang.Thread.run (Thread.java:840)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: java.io.EOFException
    at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build (SdkClientException.java:111)
    at software.amazon.awssdk.core.exception.SdkClientException.create (SdkClientException.java:47)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException (RetryableStageHelper.java:223)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException (RetryableStageHelper.java:218)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute (AsyncRetryableStage.java:182)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1 (AsyncRetryableStage.java:159)
    at java.util.concurrent.CompletableFuture.uniWhenComplete (CompletableFuture.java:863)
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire (CompletableFuture.java:841)
    at java.util.concurrent.CompletableFuture.postComplete (CompletableFuture.java:510)
    at java.util.concurrent.CompletableFuture.completeExceptionally (CompletableFuture.java:2162)
    at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0 (CompletableFutureUtils.java:79)
    at java.util.concurrent.CompletableFuture.uniWhenComplete (CompletableFuture.java:863)
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire (CompletableFuture.java:841)
    at java.util.concurrent.CompletableFuture.postComplete (CompletableFuture.java:510)
    at java.util.concurrent.CompletableFuture.completeExceptionally (CompletableFuture.java:2162)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$execute$0 (MakeAsyncHttpRequestStage.java:108)
    at java.util.concurrent.CompletableFuture.uniWhenComplete (CompletableFuture.java:863)
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire (CompletableFuture.java:841)
    at java.util.concurrent.CompletableFuture.postComplete (CompletableFuture.java:510)
    at java.util.concurrent.CompletableFuture.completeExceptionally (CompletableFuture.java:2162)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.completeResponseFuture (MakeAsyncHttpRequestStage.java:255)
    at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$3 (MakeAsyncHttpRequestStage.java:167)
    at java.util.concurrent.CompletableFuture.uniHandle (CompletableFuture.java:934)
    at java.util.concurrent.CompletableFuture$UniHandle.tryFire (CompletableFuture.java:911)
    at java.util.concurrent.CompletableFuture$Completion.run (CompletableFuture.java:482)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:840)
    Suppressed: software.amazon.awssdk.core.exception.SdkClientException: Request attempt 1 failure: Unable to execute HTTP request: java.io.EOFException
    Suppressed: software.amazon.awssdk.core.exception.SdkClientException: Request attempt 2 failure: Unable to execute HTTP request: java.io.EOFException
    Suppressed: software.amazon.awssdk.core.exception.SdkClientException: Request attempt 3 failure: Unable to execute HTTP request: java.io.EOFException
Caused by: java.io.UncheckedIOException: java.io.EOFException
    at software.amazon.awssdk.utils.FunctionalUtils.asRuntimeException (FunctionalUtils.java:180)
    at software.amazon.awssdk.utils.FunctionalUtils.lambda$safeSupplier$4 (FunctionalUtils.java:110)
    at software.amazon.awssdk.utils.FunctionalUtils.invokeSafely (FunctionalUtils.java:136)
    at software.amazon.awssdk.core.http.Crc32Validation.decompressing (Crc32Validation.java:85)
    at software.amazon.awssdk.core.http.Crc32Validation.process (Crc32Validation.java:62)
    at software.amazon.awssdk.core.http.Crc32Validation.validate (Crc32Validation.java:44)
    at software.amazon.awssdk.core.internal.handler.BaseAsyncClientHandler.lambda$new$0 (BaseAsyncClientHandler.java:66)
    at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0 (AsyncResponseHandler.java:92)
    at java.util.concurrent.CompletableFuture$UniCompose.tryFire (CompletableFuture.java:1150)
    at java.util.concurrent.CompletableFuture.postComplete (CompletableFuture.java:510)
    at java.util.concurrent.CompletableFuture.complete (CompletableFuture.java:2147)
    at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete (AsyncResponseHandler.java:135)
    at software.amazon.awssdk.core.internal.metrics.BytesReadTrackingPublisher$BytesReadTracker.onComplete (BytesReadTrackingPublisher.java:74)
    at org.apache.pekko.stream.impl.ReactiveStreamsCompliance$.tryOnComplete (ReactiveStreamsCompliance.scala:124)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$ActorOutputBoundary.complete (ActorGraphInterpreter.scala:418)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$ActorOutputBoundary.onUpstreamFinish (ActorGraphInterpreter.scala:444)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreter.processEvent (GraphInterpreter.scala:536)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreter.execute (GraphInterpreter.scala:400)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreterShell.runBatch (ActorGraphInterpreter.scala:662)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute (ActorGraphInterpreter.scala:71)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute$ (ActorGraphInterpreter.scala:67)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$RequestMore.execute (ActorGraphInterpreter.scala:299)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreterShell.processEvent (ActorGraphInterpreter.scala:637)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter.org$apache$pekko$stream$impl$fusing$ActorGraphInterpreter$$processEvent (ActorGraphInterpreter.scala:813)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse (ActorGraphInterpreter.scala:831)
    at org.apache.pekko.actor.Actor.aroundReceive (Actor.scala:547)
    at org.apache.pekko.actor.Actor.aroundReceive$ (Actor.scala:545)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter.aroundReceive (ActorGraphInterpreter.scala:729)
    at org.apache.pekko.actor.ActorCell.receiveMessage (ActorCell.scala:590)
    at org.apache.pekko.actor.ActorCell.invoke (ActorCell.scala:557)
    at org.apache.pekko.dispatch.Mailbox.processMailbox (Mailbox.scala:280)
    at org.apache.pekko.dispatch.Mailbox.run (Mailbox.scala:241)
    at org.apache.pekko.dispatch.Mailbox.exec (Mailbox.scala:253)
    at java.util.concurrent.ForkJoinTask.doExec (ForkJoinTask.java:373)
    at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec (ForkJoinPool.java:1182)
    at java.util.concurrent.ForkJoinPool.scan (ForkJoinPool.java:1655)
    at java.util.concurrent.ForkJoinPool.runWorker (ForkJoinPool.java:1622)
    at java.util.concurrent.ForkJoinWorkerThread.run (ForkJoinWorkerThread.java:165)
Caused by: java.io.EOFException
    at java.util.zip.GZIPInputStream.readUByte (GZIPInputStream.java:268)
    at java.util.zip.GZIPInputStream.readUShort (GZIPInputStream.java:258)
    at java.util.zip.GZIPInputStream.readHeader (GZIPInputStream.java:164)
    at java.util.zip.GZIPInputStream.<init> (GZIPInputStream.java:79)
    at java.util.zip.GZIPInputStream.<init> (GZIPInputStream.java:91)
    at software.amazon.awssdk.core.http.Crc32Validation.lambda$decompressing$2 (Crc32Validation.java:85)
    at software.amazon.awssdk.utils.FunctionalUtils.lambda$safeSupplier$4 (FunctionalUtils.java:108)
    at software.amazon.awssdk.utils.FunctionalUtils.invokeSafely (FunctionalUtils.java:136)
    at software.amazon.awssdk.core.http.Crc32Validation.decompressing (Crc32Validation.java:85)
    at software.amazon.awssdk.core.http.Crc32Validation.process (Crc32Validation.java:62)
    at software.amazon.awssdk.core.http.Crc32Validation.validate (Crc32Validation.java:44)
    at software.amazon.awssdk.core.internal.handler.BaseAsyncClientHandler.lambda$new$0 (BaseAsyncClientHandler.java:66)
    at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0 (AsyncResponseHandler.java:92)
    at java.util.concurrent.CompletableFuture$UniCompose.tryFire (CompletableFuture.java:1150)
    at java.util.concurrent.CompletableFuture.postComplete (CompletableFuture.java:510)
    at java.util.concurrent.CompletableFuture.complete (CompletableFuture.java:2147)
    at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete (AsyncResponseHandler.java:135)
    at software.amazon.awssdk.core.internal.metrics.BytesReadTrackingPublisher$BytesReadTracker.onComplete (BytesReadTrackingPublisher.java:74)
    at org.apache.pekko.stream.impl.ReactiveStreamsCompliance$.tryOnComplete (ReactiveStreamsCompliance.scala:124)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$ActorOutputBoundary.complete (ActorGraphInterpreter.scala:418)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$ActorOutputBoundary.onUpstreamFinish (ActorGraphInterpreter.scala:444)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreter.processEvent (GraphInterpreter.scala:536)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreter.execute (GraphInterpreter.scala:400)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreterShell.runBatch (ActorGraphInterpreter.scala:662)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute (ActorGraphInterpreter.scala:71)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute$ (ActorGraphInterpreter.scala:67)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$RequestMore.execute (ActorGraphInterpreter.scala:299)
    at org.apache.pekko.stream.impl.fusing.GraphInterpreterShell.processEvent (ActorGraphInterpreter.scala:637)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter.org$apache$pekko$stream$impl$fusing$ActorGraphInterpreter$$processEvent (ActorGraphInterpreter.scala:813)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse (ActorGraphInterpreter.scala:831)
    at org.apache.pekko.actor.Actor.aroundReceive (Actor.scala:547)
    at org.apache.pekko.actor.Actor.aroundReceive$ (Actor.scala:545)
    at org.apache.pekko.stream.impl.fusing.ActorGraphInterpreter.aroundReceive (ActorGraphInterpreter.scala:729)
    at org.apache.pekko.actor.ActorCell.receiveMessage (ActorCell.scala:590)
    at org.apache.pekko.actor.ActorCell.invoke (ActorCell.scala:557)
    at org.apache.pekko.dispatch.Mailbox.processMailbox (Mailbox.scala:280)
    at org.apache.pekko.dispatch.Mailbox.run (Mailbox.scala:241)
    at org.apache.pekko.dispatch.Mailbox.exec (Mailbox.scala:253)
    at java.util.concurrent.ForkJoinTask.doExec (ForkJoinTask.java:373)
    at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec (ForkJoinPool.java:1182)
    at java.util.concurrent.ForkJoinPool.scan (ForkJoinPool.java:1655)
    at java.util.concurrent.ForkJoinPool.runWorker (ForkJoinPool.java:1622)
```

### Reason why it's a bug
The AWS SDK expects `SdkHttpFullResponse.content` to be null/not present in the case there is no content, but this http client violates that contract.

In my sample project, you can use the built in HTTP client instead of this one, and everything works fine.

### The fix
The fix is Before [this line of code](https://github.com/pjfanning/aws-spi-pekko-http/blob/236f789f963df443a1aee0a345bc0f015c6bf162/src/main/scala/com/github/pjfanning/pekkohttpspi/RequestRunner.scala#L43) add `.filter(_.nonEmpty)`

--- a/src/main/scala/com/github/pjfanning/pekkohttpspi/RequestRunner.scala
+++ b/src/main/scala/com/github/pjfanning/pekkohttpspi/RequestRunner.scala
@@ -40,6 +40,7 @@ class RequestRunner()(implicit sys: ActorSystem, ec: ExecutionContext, mat: Mate
       handler.onHeaders(toSdkHttpFullResponse(response))
 
       val (complete, publisher) = response.entity.dataBytes
+        .filter(_.nonEmpty)
         .map(_.asByteBuffer)
         .alsoToMat(Sink.ignore)(Keep.right)
         .toMat(Sink.asPublisher(fanout = false))(Keep.both)

--- a/src/test/scala/com/github/pjfanning/pekkohttpspi/s3/TestS3.scala
+++ b/src/test/scala/com/github/pjfanning/pekkohttpspi/s3/TestS3.scala
@@ -123,6 +123,29 @@ class TestS3 extends BaseAwsClientTest[S3AsyncClient] {
       result.asUtf8String() should be(fileContent + fileContent)
     }
 
+    "upload and head request gzip file with 0 content-length" in withClient { implicit client =>
+      val bucketName = createBucket()
+      val fileName = "my-empty-file"
+      val contentLength = 0L
+      val contentType = "application/json"
+      val contentEncoding = "gzip"
+
+      client
+        .putObject(PutObjectRequest.builder().bucket(bucketName).key(fileName).contentType(contentType).contentLength(contentLength).contentEncoding(contentEncoding).build(),
+                   AsyncRequestBody.fromString("")
+        )
+        .join
+
+      val result = client
+        .headObject(HeadObjectRequest.builder().bucket(bucketName).key(fileName).build()
+        )
+        .join
+
+      result.contentLength() should be(contentLength)
+      result.contentType() should be(contentType)
+      result.contentEncoding() should be(contentEncoding)
+    }
+
   }
 
   def createBucket()(implicit client: S3AsyncClient): String = {

--- a/src/test/scala/com/github/pjfanning/pekkohttpspi/s3/TestS3.scala
+++ b/src/test/scala/com/github/pjfanning/pekkohttpspi/s3/TestS3.scala
@@ -124,21 +124,28 @@ class TestS3 extends BaseAwsClientTest[S3AsyncClient] {
     }
 
     "upload and head request gzip file with 0 content-length" in withClient { implicit client =>
-      val bucketName = createBucket()
-      val fileName = "my-empty-file"
-      val contentLength = 0L
-      val contentType = "application/json"
+      val bucketName      = createBucket()
+      val fileName        = "my-empty-file"
+      val contentLength   = 0L
+      val contentType     = "application/json"
       val contentEncoding = "gzip"
 
       client
-        .putObject(PutObjectRequest.builder().bucket(bucketName).key(fileName).contentType(contentType).contentLength(contentLength).contentEncoding(contentEncoding).build(),
-                   AsyncRequestBody.fromString("")
+        .putObject(
+          PutObjectRequest
+            .builder()
+            .bucket(bucketName)
+            .key(fileName)
+            .contentType(contentType)
+            .contentLength(contentLength)
+            .contentEncoding(contentEncoding)
+            .build(),
+          AsyncRequestBody.fromString("")
         )
         .join
 
       val result = client
-        .headObject(HeadObjectRequest.builder().bucket(bucketName).key(fileName).build()
-        )
+        .headObject(HeadObjectRequest.builder().bucket(bucketName).key(fileName).build())
         .join
 
       result.contentLength() should be(contentLength)

